src/timer.rs

```diff
@@ -288,3 +281,502 @@ where
+            context: &self.context,
+        };
+        Ok(driver)
+    }
+}
+
+impl<'a> ParallelSleepDriver<'a> {
+    /// Sleep for the given duration
+    pub async fn sleep(&self, duration: Duration<usize>) -> TockResult<()> {
+        let now = get_current_ticks()?;
+        self.context.current_time.set(now);
+        let freq = get_clock_frequency()?;
+        let alarm_instant = Self::compute_alarm_instant(duration.ms, now, freq)?;
+        let this_alarm = ActiveTimer {
+            instant: alarm_instant as u32,
+            set_at: now as u32,
+        };
+
+        let suspended_timer: Cell<Option<ActiveTimer>> = Cell::new(None);
+
+        futures::wait_until(|| {
+            let now = match get_current_ticks() {
+                Ok(value) => value,
+                Err(_) => return false,
+            };
+            if let Some(active) = self.context.active_timer.get() {
+                if left_is_later(active, this_alarm) {
+                    suspended_timer.set(Some(active));
+                    match self.activate_timer(&this_alarm) {
+                        Ok(_) => (),
+                        Err(_) => return false,
+                    };
+                }
+            } else {
+                match self.activate_timer(&this_alarm) {
+                    Ok(_) => (),
+                    Err(_) => return false,
+                };
+            }
+            if is_over(this_alarm, now as u32) {
+                if let Some(paused) = suspended_timer.get() {
+                    match self.activate_timer(&paused) {
+                        Ok(_) => (),
+                        Err(_) => return false,
+                    };
+                } else {
+                    self.context.active_timer.set(None);
+                }
+                true
+            } else {
+                false
+            }
+        })
+        .await;
+
+        Ok(())
+    }
+
+    fn activate_timer(&self, timer: &ActiveTimer) -> TockResult<()> {
+        set_alarm_at(timer.instant as usize)?;
+        let now = get_current_ticks()?;
+        if !is_over(*timer, now as u32) {
+            self.context.active_timer.set(Some(*timer));
+        } else {
+            self.wakeup_soon()?;
+        }
+        Ok(())
+    }
+
+    fn wakeup_soon(&self) -> TockResult<()> {
+        let mut i = 0;
+        self.context.active_timer.set(None);
+
+        loop {
+            let now = get_current_ticks()?;
+
+            let next_timer = ActiveTimer {
+                instant: now as u32 + i,
+                set_at: now as u32,
+            };
+            set_alarm_at(next_timer.instant as usize)?;
+            let now = get_current_ticks()?;
+            if !is_over(next_timer, now as u32) {
+                break;
+            } else {
+                match stop_alarm_at(next_timer.instant as usize) {
+                    Ok(_) => (),
+                    Err(_) => (),
+                }
+                i += 1;
```

Comment:
Shouldn't the loop be `for i in 0..` instead of this manual incrementing?

Reply:
Fixed.